### ISSUE TITLE
SCRFD vs TinaFace vs Retinaface 

### ISSUE THREAD

**Issue by cxz1418** [deepinsight/insightface#1639]
HI  I'm analyzing the figures related to the face-detector
,but there is something I don't understand, so I'm posting it.

I am comparing the performance of SCRFD 
with retinaface and Tinaface, but it came out too low than the number in SCRFD paper.

Below are the performance figures for each paper.

What am I missing?

## SCRFD paper
![image](https://user-images.githubusercontent.com/11434363/124897675-7ebc3580-e019-11eb-94eb-e059f488030b.png)

## Tinaface paper
![image](https://user-images.githubusercontent.com/11434363/124897840-a4493f00-e019-11eb-8072-d7500b5879e1.png)

## Retinaface paper
![image](https://user-images.githubusercontent.com/11434363/124897909-b1662e00-e019-11eb-95f2-d3311f253d48.png)

**Comment by nttstar** [comment]
SCRFD reports the performance under 640x480 input size.

**Comment by xsacha** [comment]
Retinaface performs better at original scale and SCRFD performs worse at anything other than trained resolution (i.e. 640px) in my experience. The result is retinaface performs better when constrained by a minimum face pixel size, where it can run below trained resolution (i.e. 640px) and get the same accuracy.

Also, I found in practice that retinaface was in fact faster for batched inputs and the same speed for single batch. I do cache bounding boxes and such just as SCRFD does, which may be the difference in reported timings. Basically if you do all the ticks SCRFD does, you get the same or better inference time in the competing algorithms anyway.

**Comment by nttstar** [comment]
You can not constrain the system by a minimum face pixel size as in practice we often resize the input image to a smaller size and finally enlarge the result face boxes to their original scale.

For inference time, we just report the network forwarding time, without post-processing such as the bboxes cache you mentioned, which is not consistent with your results.

**Comment by xsacha** [comment]
Resizing the image, for Retinaface, is how you constrain the min face size. It doesn't appear to work for SCRFD however as you lose accuracy.

**Comment by nttstar** [comment]
Inference time of very small models have very little difference on modern GPUs.
You may want to test it on CPU single-thread or ARM devices.

**Comment by xsacha** [comment]
I was encouraged initially by the times and promise of higher accuracy but as I tested it later with larger batch sizes and datasets I found it did not perform as fast unfortunately.
For the OP @cxz1418, it is something to keep in mind that this table - these statistics - based on Widerface val is not a reliable metric.

**Comment by nttstar** [comment]
It will be affected by many factors in real-world applications. Ten people may have ten conclusions.

**Comment by nttstar** [comment]
And I can not imagine that a similar input size of wider-face test-est like 1024x768, used in your system.

**Comment by xsacha** [comment]
Another thing to be mindful of is the peak memory usage that is not shown here. The depthwise conv performs badly within PyTorch for memory usage.
Consistent profiles are better for production.

Following comparison is PASCAL dataset with no post-processing.
Retina MNet 0.5G (3.3GBp and fairly consistent profile):
![image](https://user-images.githubusercontent.com/61218/125243851-f5b34000-e331-11eb-8f00-60d775232e3c.png)

SCRFD0.5G (4.8GBp and inconsistent profile):
![image](https://user-images.githubusercontent.com/61218/125244094-3f038f80-e332-11eb-9111-851d3d5455f3.png)

**Comment by nttstar** [comment]
Yes, that's why I said 'ten people may have ten conclusions', many issues must be considered for real-world deployment.

**Comment by fisakhan** [comment]
@xsacha Can you please help me converting SCRFD to C++ for inference only?

**Comment by MyraBaba** [comment]
@nttstar 

For recognition as you know the alignment is important. 

For best accuracy which one is more powerful retinaface or SCRFD ? Or do we need extra 1k3d68 / 2d106det landmark processing for best alignment ?

thanks

**Comment by pdd-99** [comment]
> 

Would you mind sharing which retinaface variant model you are using? Is the model publicity available or did you train it yourself?
@xsacha

**Comment by pdd-99** [comment]
@xsacha Would you mind sharing which retinaface variant you are using? Is it publicity available or did you train it yourself?
